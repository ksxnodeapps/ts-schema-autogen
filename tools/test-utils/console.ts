import { ConsoleInstance } from 'simple-fake-console'
import { Console } from '@ts-schema-autogen/types'

export class FakeConsole implements Console.Mod {
  public readonly core = new ConsoleInstance()
  public readonly info = jest.fn(this.core.info)
  public readonly error = jest.fn(this.core.error)
}
