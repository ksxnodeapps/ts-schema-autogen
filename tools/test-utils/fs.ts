import { StringPathFileSystem } from 'simple-fake-fs'
import { FSX } from '@ts-schema-autogen/types'

export class FakeFileSystem extends StringPathFileSystem implements FSX.Mod {
  private mkfn<Xs extends any[], Y> (fn: (...xs: Xs) => Y) {
    return jest.fn(async (...xs: Xs) => fn.apply(this, xs))
  }

  public readonly stat = this.mkfn(this.statSync)
  public readonly readdir = this.mkfn(this.readdirSync)
  public readonly readFile = this.mkfn(this.readFileSync)
  public readonly outputFile = this.mkfn(this.outputFileSync)
  public readonly writeFile = this.mkfn(this.writeFileSync)
  public readonly remove = jest.fn()
}
