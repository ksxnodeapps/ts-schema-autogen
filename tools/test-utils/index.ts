export * from './console'
export * from './fs'
export * from './path'
export * from './print-result'
export * from './process'
export * from './typescript-json-schema'
