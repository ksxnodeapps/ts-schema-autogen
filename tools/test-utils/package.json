{
  "name": "@tools/test-utils",
  "version": "0.0.0",
  "private": true,
  "dependencies": {
    "simple-fake-fs": "^0.0.8",
    "simple-fake-console": "^0.1.6",
    "simple-fake-path": "^0.0.2",
    "@ts-schema-autogen/status": "file:../../packages/status",
    "@ts-schema-autogen/types": "file:../../packages/types"
  }
}
