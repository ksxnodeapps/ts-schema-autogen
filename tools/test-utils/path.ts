import { Path } from '@ts-schema-autogen/types'

export class FakePath implements Path.Mod {
  constructor (private readonly cwd: string) {}

  public readonly join = (...paths: string[]): string => {
    if (!paths.length) return '.'
    const [head, ...rest] = paths
    const tail = this.join(...rest)
    if (head === '.') return tail
    if (tail === '.') return head
    if (tail.startsWith('../')) return this.join(this.dirname(head), tail.slice('../'.length))
    const left = head.endsWith('/') ? head.slice(0, -1) : head
    const right = tail.startsWith('/') ? tail.slice(1) : tail
    return left + '/' + right
  }

  public readonly resolve = (...paths: string[]): string => {
    if (!paths.length) return '.'
    const [head, ...rest] = paths
    const tail = this.resolve(...rest)
    if (tail.startsWith('/')) return tail
    return this.join(this.cwd, head, tail)
  }

  public readonly dirname = (path: string): string => {
    if (path === '' || path === '.' || path === '/') return path
    const segments = path.split('/').slice(0, -1)
    if (!segments.length) return '.'
    return this.join(...segments)
  }

  public readonly basename = (path: string): string => {
    const segments = path.split('/')
    return segments[segments.length - 1]
  }
}
