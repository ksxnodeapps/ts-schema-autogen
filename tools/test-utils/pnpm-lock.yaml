dependencies:
  '@ts-schema-autogen/types': 'link:../../packages/types'
  simple-fake-fs: 0.0.4
lockfileVersion: 5.1
packages:
  /@types/node/13.5.0:
    dev: false
    resolution:
      integrity: sha512-Onhn+z72D2O2Pb2ql2xukJ55rglumsVo1H6Fmyi8mlU9SvKdBk/pUSUAiBY/d9bAOF7VVWajX3sths/+g6ZiAQ==
  /simple-fake-fs/0.0.4:
    dependencies:
      '@types/node': 13.5.0
      tslib: 1.10.0
    dev: false
    resolution:
      integrity: sha512-GUwEIYXI6oSJdMZkP1hK3x084z6HX6uyHMrs3erJXEzdyMIrJ7HJEjlbtMrs6zsELvgx0wr7Y9zTK61+igHC8A==
  /tslib/1.10.0:
    dev: false
    resolution:
      integrity: sha512-qOebF53frne81cf0S9B41ByenJ3/IuH8yJKngAX35CmiZySA0khhkovshKK+jGCaMnVomla7gVlIcc3EvKPbTQ==
specifiers:
  '@ts-schema-autogen/types': ../../packages/types
  simple-fake-fs: ^0.0.4
