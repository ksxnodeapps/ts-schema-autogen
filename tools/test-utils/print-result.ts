import { inspect } from 'util'
import { Success, Failure } from '@ts-schema-autogen/status'

export function printFailure (result: Failure<unknown>): string {
  let text = '\n'
  result.print((...args) => text += args.join(' ') + '\n')
  return text
}

export const printResult = (result: Success<unknown> | Failure<unknown>): string =>
  result.code ? printFailure(result) : inspect(result)
