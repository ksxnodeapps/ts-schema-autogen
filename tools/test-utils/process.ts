import { Process } from '@ts-schema-autogen/types'

export class FakeProcess implements Process.Mod {
  constructor (private readonly workDir: string) {}
  public readonly cwd = jest.fn(() => this.workDir)
}
