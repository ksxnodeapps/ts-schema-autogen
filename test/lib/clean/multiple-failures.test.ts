import { inspect } from 'util'
import createFsTree from '../../fixtures/fs-tree/multiple-failures'
import TJS from '../../fixtures/tjs'
import { FakeFileSystem, FakePath, printResult } from '@tools/test-utils'

import {
  Status,
  MultipleFailures,
  clean,
  createJsonConfigParser,
  createYamlConfigParser
} from '@ts-schema-autogen/lib'

const getParsers = () => Promise.all([
  createJsonConfigParser('JSON Parser'),
  createYamlConfigParser('YAML Parser')
])

class MockedFileSystem extends FakeFileSystem {
  public readonly unlink = jest.fn(
    (path: string) => path.endsWith('.removable')
      ? Promise.resolve()
      : Promise.reject(new RemovalError(path))
  )
}

class RemovalError extends Error {
  constructor (path: string) {
    super(`File or directory ${inspect(path)} cannot be deleted`)
  }

  public readonly name = this.constructor.name

  public [inspect.custom] () {
    return `${this.name}: ${this.message}`
  }
}

async function setup (configFiles: readonly string[]) {
  const fsx = new MockedFileSystem('/', await createFsTree())
  const path = new FakePath()
  const tjs = new TJS()
  const result = await clean({
    configFiles,
    fsx,
    path,
    parsers: await getParsers()
  })
  return { fsx, path, tjs, result }
}

const configFiles = [
  'circular-reference/1/0.json',
  'circular-reference/2/0.json',
  'circular-reference/3/0.json',
  'file-reading-failure/single.json',
  'file-reading-failure/multiple.json',
  'file-removal-failure/container/foo.json',
  'file-removal-failure/container/bar.json',
  'file-removal-failure/container/baz.json',
  'output-file-conflict/obvious-conflicts/foo.json',
  'output-file-conflict/obvious-conflicts/bar.json',
  'text-parsing-failure/invalid-config',
  'valid-config-only/yaml/single-symbol/single-output/output-filename/.schema.autogen.yaml',
  'valid-config-only/yaml/single-symbol/single-output/output-descriptor/.schema.autogen.yaml',
  'valid-config-only/yaml/single-symbol/multiple-output/output-filename/.schema.autogen.yaml'
]

it('returns result matching snapshot', async () => {
  const { result } = await setup(configFiles)
  expect(result).toMatchSnapshot()
})

it('returns a MultipleFailures', async () => {
  const { result } = await setup(configFiles)
  expect(result).toBeInstanceOf(MultipleFailures)
})

it('returns a result containing expected properties', async () => {
  const { result } = await setup(configFiles)
  expect(result).toMatchObject({
    code: Status.MultipleFailures,
    error: expect.any(Array)
  })
})

it('error messages', async () => {
  const { result } = await setup(configFiles)
  expect(printResult(result)).toMatchSnapshot()
})

it('status code', async () => {
  const { result } = await setup(configFiles)
  expect(result.getStatusCode()).toBe(Status.MultipleFailures)
})
