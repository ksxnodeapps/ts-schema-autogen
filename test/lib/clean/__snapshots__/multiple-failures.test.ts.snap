// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error messages 1`] = `
"
MultipleFailures
    CircularReference
        circular-reference/1/0.json
    CircularReference
        circular-reference/2/0.json
        circular-reference/2/1.json
    CircularReference
        circular-reference/3/0.json
        circular-reference/3/1.json
        circular-reference/3/2.json
    FileReadingFailure
        path: file-reading-failure/not-exist.json
        error:
            ENOENT {
              syscall: 'read',
              path: [ 'file-reading-failure', 'not-exist.json' ],
              errno: -2,
              code: 'ENOENT'
            }
    FileReadingFailure
        path: file-reading-failure/not-exist-0.json
        error:
            ENOENT {
              syscall: 'read',
              path: [ 'file-reading-failure', 'not-exist-0.json' ],
              errno: -2,
              code: 'ENOENT'
            }
    FileRemovalFailure
        path: file-removal-failure/container/bar.unremovable
        error:
            RemovalError: File or directory 'file-removal-failure/container/bar.unremovable' cannot be deleted
    FileRemovalFailure
        path: file-removal-failure/container/baz.unremovable
        error:
            RemovalError: File or directory 'file-removal-failure/container/baz.unremovable' cannot be deleted
    FileRemovalFailure
        path: output-file-conflict/obvious-conflicts/shared-output.json
        error:
            RemovalError: File or directory 'output-file-conflict/obvious-conflicts/shared-output.json' cannot be deleted
    FileRemovalFailure
        path: output-file-conflict/obvious-conflicts/shared-output.json
        error:
            RemovalError: File or directory 'output-file-conflict/obvious-conflicts/shared-output.json' cannot be deleted
    TextParsingFailure
        JSON Parser
            SyntaxError: Unexpected token e in JSON at position 1
        YAML Parser
            YAMLException: incomplete explicit mapping pair; a key node is missed; or followed by a non-tabulated empty line in \\"text-parsing-failure/invalid-config\\" at line 1, column 11:
                neither: a: valid: yaml: nor: a: valid: json:
                          ^
    FileRemovalFailure
        path: valid-config-only/yaml/single-symbol/single-output/output-filename/output.schema.json
        error:
            RemovalError: File or directory 'valid-config-only/yaml/single-symbol/single-output/output-filename/output.schema.json' cannot be deleted
    FileRemovalFailure
        path: valid-config-only/yaml/single-symbol/single-output/output-descriptor/output.schema.json
        error:
            RemovalError: File or directory 'valid-config-only/yaml/single-symbol/single-output/output-descriptor/output.schema.json' cannot be deleted
    FileRemovalFailure
        path: valid-config-only/yaml/single-symbol/multiple-output/output-filename/abc.schema.json
        error:
            RemovalError: File or directory 'valid-config-only/yaml/single-symbol/multiple-output/output-filename/abc.schema.json' cannot be deleted
    FileRemovalFailure
        path: valid-config-only/yaml/single-symbol/multiple-output/output-filename/def.schema.json
        error:
            RemovalError: File or directory 'valid-config-only/yaml/single-symbol/multiple-output/output-filename/def.schema.json' cannot be deleted
    FileRemovalFailure
        path: valid-config-only/yaml/single-symbol/multiple-output/output-filename/ghi.schema.json
        error:
            RemovalError: File or directory 'valid-config-only/yaml/single-symbol/multiple-output/output-filename/ghi.schema.json' cannot be deleted
"
`;

exports[`returns result matching snapshot 1`] = `
MultipleFailures {
  "code": 2,
  "error": Array [
    CircularReference {
      "code": 8,
      "error": Array [
        "circular-reference/1/0.json",
      ],
      "value": undefined,
    },
    CircularReference {
      "code": 8,
      "error": Array [
        "circular-reference/2/0.json",
        "circular-reference/2/1.json",
      ],
      "value": undefined,
    },
    CircularReference {
      "code": 8,
      "error": Array [
        "circular-reference/3/0.json",
        "circular-reference/3/1.json",
        "circular-reference/3/2.json",
      ],
      "value": undefined,
    },
    FileReadingFailure {
      "code": 5,
      "error": ENOENT {
        "code": "ENOENT",
        "errno": -2,
        "path": Array [
          "file-reading-failure",
          "not-exist.json",
        ],
        "syscall": "read",
      },
      "path": "file-reading-failure/not-exist.json",
      "value": undefined,
    },
    FileReadingFailure {
      "code": 5,
      "error": ENOENT {
        "code": "ENOENT",
        "errno": -2,
        "path": Array [
          "file-reading-failure",
          "not-exist-0.json",
        ],
        "syscall": "read",
      },
      "path": "file-reading-failure/not-exist-0.json",
      "value": undefined,
    },
    FileRemovalFailure {
      "code": 6,
      "error": [RemovalError: File or directory 'file-removal-failure/container/bar.unremovable' cannot be deleted],
      "path": "file-removal-failure/container/bar.unremovable",
      "value": undefined,
    },
    FileRemovalFailure {
      "code": 6,
      "error": [RemovalError: File or directory 'file-removal-failure/container/baz.unremovable' cannot be deleted],
      "path": "file-removal-failure/container/baz.unremovable",
      "value": undefined,
    },
    FileRemovalFailure {
      "code": 6,
      "error": [RemovalError: File or directory 'output-file-conflict/obvious-conflicts/shared-output.json' cannot be deleted],
      "path": "output-file-conflict/obvious-conflicts/shared-output.json",
      "value": undefined,
    },
    FileRemovalFailure {
      "code": 6,
      "error": [RemovalError: File or directory 'output-file-conflict/obvious-conflicts/shared-output.json' cannot be deleted],
      "path": "output-file-conflict/obvious-conflicts/shared-output.json",
      "value": undefined,
    },
    TextParsingFailure {
      "code": 7,
      "error": Array [
        Object {
          "error": [SyntaxError: Unexpected token e in JSON at position 1],
          "parser": Object {
            "name": "JSON Parser",
            "parseConfigText": [Function],
            "testFileName": [Function],
          },
        },
        Object {
          "error": [YAMLException: incomplete explicit mapping pair; a key node is missed; or followed by a non-tabulated empty line in "text-parsing-failure/invalid-config" at line 1, column 11:
    neither: a: valid: yaml: nor: a: valid: json:
              ^],
          "parser": Object {
            "name": "YAML Parser",
            "parseConfigText": [Function],
            "testFileName": [Function],
          },
        },
      ],
      "value": undefined,
    },
    FileRemovalFailure {
      "code": 6,
      "error": [RemovalError: File or directory 'valid-config-only/yaml/single-symbol/single-output/output-filename/output.schema.json' cannot be deleted],
      "path": "valid-config-only/yaml/single-symbol/single-output/output-filename/output.schema.json",
      "value": undefined,
    },
    FileRemovalFailure {
      "code": 6,
      "error": [RemovalError: File or directory 'valid-config-only/yaml/single-symbol/single-output/output-descriptor/output.schema.json' cannot be deleted],
      "path": "valid-config-only/yaml/single-symbol/single-output/output-descriptor/output.schema.json",
      "value": undefined,
    },
    FileRemovalFailure {
      "code": 6,
      "error": [RemovalError: File or directory 'valid-config-only/yaml/single-symbol/multiple-output/output-filename/abc.schema.json' cannot be deleted],
      "path": "valid-config-only/yaml/single-symbol/multiple-output/output-filename/abc.schema.json",
      "value": undefined,
    },
    FileRemovalFailure {
      "code": 6,
      "error": [RemovalError: File or directory 'valid-config-only/yaml/single-symbol/multiple-output/output-filename/def.schema.json' cannot be deleted],
      "path": "valid-config-only/yaml/single-symbol/multiple-output/output-filename/def.schema.json",
      "value": undefined,
    },
    FileRemovalFailure {
      "code": 6,
      "error": [RemovalError: File or directory 'valid-config-only/yaml/single-symbol/multiple-output/output-filename/ghi.schema.json' cannot be deleted],
      "path": "valid-config-only/yaml/single-symbol/multiple-output/output-filename/ghi.schema.json",
      "value": undefined,
    },
  ],
  "value": undefined,
}
`;
